Replace our custom compact tables with the Table component from PF-4 react

KKoukiou opened the pull request:
Storage Volumes
![Screen Shot 2019-08-30 at 15 03 04](https://user-images.githubusercontent.com/14921356/64022891-6fbfee80-cb37-11e9-94d3-8b3eeea09cf8.png)


NICS
![Screen Shot 2019-08-30 at 15 02 46](https://user-images.githubusercontent.com/14921356/64022815-4a32e500-cb37-11e9-9860-b34397af351d.png)

Disks
![Screen Shot 2019-08-30 at 15 02 41](https://user-images.githubusercontent.com/14921356/64022816-4a32e500-cb37-11e9-9507-2d4cb361c0b8.png)

marusak commented on pkg/machines/components/storagePools/storagePoolVolumesTab.jsx:
```diff
@@ -33,37 +37,45 @@ export class StoragePoolVolumesTab extends React.Component {
 
-    deleteErrorHandler(deleteError, deleteErrorDetail) {
-        this.setState({ deleteError, deleteErrorDetail });
+    componentWillReceiveProps(nextProps) {
```
isn't `componentWillReceiveProps` being obsoleted?

KKoukiou replied:
Yes, needs fix.

marusak commented on pkg/machines/machines.less:
```diff
@@ -5,10 +5,6 @@
 @import (less) "../../node_modules/c3/c3.css";
 
 // For patternfly-react and @patternfly/react-console
-@import "~xterm/lib/xterm.css";
```
Why you dropped this in this commit? Is it somehow related?

KKoukiou replied:
We don't seem to need it, the component import what it needs.
But it should be separate commit indeed.

KKoukiou commented:
Tests need some extra attention.

KKoukiou commented:
The new loaders are 400KB in total together.
```
[kkoukiou@sequoia cockpit-bak]$ du -sh node_modules/file-loader/
132K	node_modules/file-loader/
[kkoukiou@sequoia cockpit-bak]$ du -sh node_modules/url-loader/
276K	node_modules/url-loader/
```

marusak commented:
[fedora-30](https://209.132.184.41:8493/logs/pull-12573-20190829-131414-6ad836c3-cockpit-project-cockpit--fedora-30/log.html) one flake and [edge](https://209.132.184.41:8493/logs/pull-12573-20190829-131414-6ad836c3-cockpit-project-cockpit--fedora-30-selenium-edge/log.html) rather red. Once more for comparison.

martinpitt commented:
[edge](https://209.132.184.41:8493/logs/pull-12573-20190830-044646-6ad836c3-cockpit-project-cockpit--fedora-30-selenium-edge/log.html) is still very red, and [screenshots](https://209.132.184.41:8493/logs/pull-12573-20190830-044646-6ad836c3-cockpit-project-cockpit--fedora-30-selenium-edge/screenshot-testNetworkInfo-325523.png) show an oops -- so this is clearly a bug somewhere.

martinpitt commented on pkg/machines/components/storagePools/storagePoolVolumesTab.jsx:
```diff
@@ -72,50 +85,53 @@ export class StoragePoolVolumesTab extends React.Component {
+                    res.cells = columns;
+                    res.selected = volume.selected;
+
+                    return res;
```
`return { cells: columns, selected: volume.selected }` ?

martinpitt commented on pkg/machines/machines.less:
```diff
@@ -6,9 +6,6 @@
 
 // For patternfly-react and @patternfly/react-console
 @import "~xterm/lib/xterm.css";
```
Seems you wanted to remove this in a separate commit instead, but there is none?

martinpitt replied:
Still outstanding?

KKoukiou replied:
Ah sorry forgot to answer. I don't think this should be there, but I don't see why this removal should be part of this PR either. Does not get anyhow affected by the PF4 things.

martinpitt commented:
I tested this and at least to me these tables look just fine, thanks! On the Networks page table it's a bit odd that there's a single "Overview" tab in the expander, but we've had that on our cockpit Listing as well. I added two screenshots to the description.

KKoukiou commented:
> I tested this and at least to me these tables look just fine, thanks! On the Networks page table it's a bit odd that there's a single "Overview" tab in the expander, but we've had that on our cockpit Listing as well. I added two screenshots to the description.

This PR did not affect the Networks Page, it's only about the list which were using the 'compact' class.
That only Disk, NICs and Storage Volumes. I updated the screenshots accordingly.

martinpitt commented on package.json:
```diff
@@ -3,13 +3,17 @@
   "dependencies": {
+    "@patternfly/patternfly": "2.26.1",
     "@patternfly/react-console": "1.12.2",
+    "@patternfly/react-core": "^3.86.2",
```
Please drop the ^, also further down for all the other modules.

martinpitt commented on webpack.config.js:
```diff
@@ -414,6 +414,14 @@ module.exports = {
             },
+            {
+                test: /\.(eot|woff|woff2|svg|ttf)([\?]?.*)$/,
+                loader: 'file-loader',
```
I still wonder about this bit. Are these just tiny fonts with one glyph or two for specialized widgets? But where would a table even use these? We really don't want to load complete fonts that way, we should instead serve them from cockpit/static in cockpit-ws, where they can be cached properly. What happens if you drop this?

KKoukiou replied:
We don't seem to need it anymore. It previous versions of PF-react it webpack was dailing without this. Dropped it.

martinpitt commented on webpack.config.js:
```diff
@@ -414,6 +414,14 @@ module.exports = {
+            },
+            {
+                test: /\.(jpg)$/i,
+                loader: 'url-loader',
```
Likewise, do tables use any .jpgs for styling actually? Do the tables here look any different without this extra loader?

KKoukiou replied:
Without this webpack fails with errors like the following.
```
    ERROR in ./node_modules/@patternfly/react-styles/css/assets/images/pfbg_576.jpg 1:0
    Module parse failed: Unexpected character '�' (1:0)
    You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
    (Source code omitted for this binary file)
     @ ./node_modules/@patternfly/react-styles/css/components/BackgroundImage/background-image.css (./node_modules/css-loader?minimize=&root=/home/kkoukiou/repos/cockpit-bak/pkg/lib!./node_modules/@patternfly/react-styles/css/components/BackgroundImage/background-image.css) 7:274-317
    
    ERROR in ./node_modules/@patternfly/react-styles/css/assets/images/pfbg_576@2x.jpg 1:0
    Module parse failed: Unexpected character '�' (1:0)
    You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
    (Source code omitted for this binary file)
     @ ./node_modules/@patternfly/react-styles/css/components/BackgroundImage/background-image.css (./node_modules/css-loader?minimize=&root=/home/kkoukiou/repos/cockpit-bak/pkg/lib!./node_modules/@patternfly/react-styles/css/components/BackgroundImage/background-image.css) 7:388-434
    
    ERROR in ./node_modules/@patternfly/react-styles/css/assets/images/pfbg_768.jpg 1:0
    Module parse failed: Unexpected character '�' (1:0)
    You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
    (Source code omitted for this binary file)
     @ ./node_modules/@patternfly/react-styles/css/components/BackgroundImage/background-image.css (./node_modules/css-loader?minimize=&root=/home/kkoukiou/repos/cockpit-bak/pkg/lib!./node_modules/@patternfly/react-styles/css/components/BackgroundImage/background-image.css) 7:506-549
    
    ERROR in ./node_modules/@patternfly/react-styles/css/assets/images/pfbg_768@2x.jpg 1:0
    Module parse failed: Unexpected character '�' (1:0)
    You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
    (Source code omitted for this binary file)
     @ ./node_modules/@patternfly/react-styles/css/components/BackgroundImage/background-image.css (./node_modules/css-loader?minimize=&root=/home/kkoukiou/repos/cockpit-bak/pkg/lib!./node_modules/@patternfly/react-styles/css/components/BackgroundImage/background-image.css) 7:624-670
    

```

martinpitt commented on pkg/lib/cockpit-components-listing.jsx:
```diff
@@ -334,8 +333,6 @@ export const Listing = (props) => {
     if (props.fullWidth)
         bodyClasses.push("listing-ct-wide");
-    if (props.compact)
-        bodyClasses.push("listing-ct-compact");
```
This class should then also be cleaned up from pkg/lib/listing.less. This can be a follow-up (until I find something else). Also, dropping the compact feature from the Listing component should be a separate commit, so that it's easier to cherry-pick into external branches.

KKoukiou commented:
> Aside from a small nitpick, the code changes look good so far. But I want to point out that this is quite a dramatic increase in size. On master:
> 
> ```
> ❱❱❱ du -hsc dist/machines/machines.{css,js}
> 80K	dist/machines/machines.css
> 7,7M	dist/machines/machines.js
> 7,8M	total
> ```
> 
> and with this PR:
> 
> ```
> ❱❱❱ du -hsc dist/machines/machines.{css,js}
> 2,6M	dist/machines/machines.css
> 16M	dist/machines/machines.js
> 18M	total
> ```
> 
> Is there any way we can reduce the impact of this? By only importing some subset, or fixing our webpack to throw out stuff that isn't being used? It's hard to imagine that the new table needs that much CSS and 10 MB of extra JavaScript code. In fact, it's hard to imagine that the _entirety_ of PF4 uses that much -- this really smells like some embedded resources or tons of new bundled libraries which shouldn't be there?
> 
> Also, oddly enough dist/playground/react-patterns.js balloons from 1 MB to 5.6 MB. All other changes apply to pkg/machines/ only, so could the new url-loader trigger this somehow?

Without the url import for the jpg images the css size is decreased, but of course the JS remains the 
same.
See: +                        options: { url: false }
in css loader.

[kkoukiou@sequoia cockpit-bak]$ du -hsc dist/machines/machines.{css,js}
500K	dist/machines/machines.css
16M	dist/machines/machines.js
16M	total

So this needs more investigated.

KKoukiou commented:
> > Aside from a small nitpick, the code changes look good so far. But I want to point out that this is quite a dramatic increase in size. On master:
> > ```
> > ❱❱❱ du -hsc dist/machines/machines.{css,js}
> > 80K	dist/machines/machines.css
> > 7,7M	dist/machines/machines.js
> > 7,8M	total
> > ```
> > 
> > 
> > and with this PR:
> > ```
> > ❱❱❱ du -hsc dist/machines/machines.{css,js}
> > 2,6M	dist/machines/machines.css
> > 16M	dist/machines/machines.js
> > 18M	total
> > ```
> > 
> > 
> > Is there any way we can reduce the impact of this? By only importing some subset, or fixing our webpack to throw out stuff that isn't being used? It's hard to imagine that the new table needs that much CSS and 10 MB of extra JavaScript code. In fact, it's hard to imagine that the _entirety_ of PF4 uses that much -- this really smells like some embedded resources or tons of new bundled libraries which shouldn't be there?
> > Also, oddly enough dist/playground/react-patterns.js balloons from 1 MB to 5.6 MB. All other changes apply to pkg/machines/ only, so could the new url-loader trigger this somehow?
> 
> Without the url import for the jpg images the css size is decreased, but of course the JS remains the
> same.
> See: + options: { url: false }
> in css loader.
> 
> [kkoukiou@sequoia cockpit-bak]$ du -hsc dist/machines/machines.{css,js}
> 500K dist/machines/machines.css
> 16M dist/machines/machines.js
> 16M total
> 
> So this needs more investigation.

@martinpitt  Regarding the increase in the JS size, maybe it's just the react-table?
I see:
```
[kkoukiou@sequoia cockpit-bak]$ du -hsc node_modules/@patternfly/react-table/src/components/Table/
4.8M	node_modules/@patternfly/react-table/src/components/Table/
4.8M	total
```
Almost 5M, so this can be it?

martinpitt commented:
> Also, oddly enough dist/playground/react-patterns.js balloons from 1 MB to 5.6 MB.

This was a red herring, I suppose I caught a built tree where dist/ was not built in debug mode (i. e. no minifying etc.)

The 80 kB → 500 kB machines.css diff is indeed all PF CSS, no encoded JPGs or anything like that. But it looks like most or all of PF4, not just the tables -- e. g. there's pf-c-wizard, pf-c-about-modal-box¸ pf-c-login, and all the other stuff that we are *never* going to need on the Machines page.

This is also the case for machines.js -- just check this:

    grep -o 'node_modules/@patternfly/[^" ]*' dist/machines/machines.js | sort -u

This JS also has tons of icons encoded in JS, which is terribly inefficient. For example:
```
/*!****************************************************************************!*\
  !*** ./node_modules/@patternfly/react-icons/dist/esm/icons/shower-icon.js ***!
  \****************************************************************************/
/*! exports provided: ShowerIconConfig, default */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
__webpack_require__.r(__webpack_exports__);
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "ShowerIconConfig", function() { return ShowerIconConfig; });
/* harmony import */ var _createIcon__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ../createIcon */ "./node_modules/@patternfly/react-icons/dist/esm/createIcon.js");
/* This file is generated by createIcons.js any changes will be lost. */

const ShowerIconConfig = {
  name: 'ShowerIcon',
  height: 512,
  width: 512,
  svgPath: 'M389.66 135.6L231.6 293.66c-9.37 9.37-24.57 9.37-33.94 0l-11.32-11.32c-9.37-9.37-9.37-24.57 0-33.94l.11-.11c-34.03-40.21-35.16-98.94-3.39-140.38-11.97-7.55-26.14-11.91-41.3-11.91C98.88 96 64 130.88 64 173.76V480H0V173.76C0 95.59 63.59 32 141.76 32c36.93 0 70.61 14.2 95.86 37.42 35.9-11.51 76.5-4.5 106.67 21.03l.11-.11c9.37-9.37 24.57-9.37 33.94 0l11.32 11.32c9.37 9.37 9.37 24.57 0 33.94zM384 208c0 8.837-7.163 16-16 16s-16-7.163-16-16 7.163-16 16-16 16 7.163 16 16zm32 0c0-8.837 7.163-16 16-16s16 7.163 16 16-7.163 16-16 16-16-7.163-16-16zm96 0c0 8.837-7.163 16-16 16s-16-7.163-16-16 7.163-16 16-16 16 7.163 16 16zm-160 32c0 8.837-7.163 16-16 16s-16-7.163-16-16 7.163-16 16-16 16 7.163 16 16zm48-16c8.837 0 16 7.163 16 16s-7.163 16-16 16-16-7.163-16-16 7.163-16 16-16zm80 16c0 8.837-7.163 16-16 16s-16-7.163-16-16 7.163-16 16-16 16 7.163 16 16zm-160 32c0 8.837-7.163 16-16 16s-16-7.163-16-16 7.163-16 16-16 16 7.163 16 16zm32 0c0-8.837 7.163-16 16-16s16 7.163 16 16-7.163 16-16 16-16-7.163-16-16zm96 0c0 8.837-7.163 16-16 16s-16-7.163-16-16 7.163-16 16-16 16 7.163 16 16zm-128 32c0-8.837 7.163-16 16-16s16 7.163 16 16-7.163 16-16 16-16-7.163-16-16zm96 0c0 8.837-7.163 16-16 16s-16-7.163-16-16 7.163-16 16-16 16 7.163 16 16zm-96 32c0 8.837-7.163 16-16 16s-16-7.163-16-16 7.163-16 16-16 16 7.163 16 16zm64 0c0 8.837-7.163 16-16 16s-16-7.163-16-16 7.163-16 16-16 16 7.163 16 16zm-32 32c0 8.837-7.163 16-16 16s-16-7.163-16-16 7.163-16 16-16 16 7.163 16 16zm-32 32c0 8.837-7.163 16-16 16s-16-7.163-16-16 7.163-16 16-16 16 7.163 16 16z',
  yOffset: '',
  xOffset: '',
  transform: ''
};
```

It seems that this isn't something which  we can easily untangle as the consumer. This needs to be fixed upstream in their build systems, so that the components really can be used standalone without pulling in the entirety of PF4 (with all icons etc.) into every page.

As a next step I propose to create a really tiny demo project that just instantiates a `<Table>` and builds it. It can be a trivial .html page that just directly `<script>`-includes the js, just to demonstrate that the table works. There we can check the size of the generated files, and refer to it in an upstream bug report.

martinpitt commented:
It also seems that merely npm-installing that particular component slurps in the entire PatternFly as a submodule anyway:

```
17M	node_modules/@patternfly/react-table/node_modules/@patternfly/patternfly/
```

KKoukiou commented:
@martinpitt Also I was trying the production build on master vs my PR and the difference is not that bad.

```
THIS PR
[kkoukiou@sequoia cockpit-bak]$ du -sh dist/machines/*
4.0K	dist/machines/index.html
192K	dist/machines/machines.css
228K	dist/machines/machines.css.map
1.5M	dist/machines/machines.min.js
6.0M	dist/machines/machines.min.js.map
148K	dist/machines/Makefile.deps
4.0K	dist/machines/manifest.json
0	dist/machines/stamp
8.0K	dist/machines/test-machines.css
12K	dist/machines/test-machines.css.map
4.0K	dist/machines/test-machines.html
80K	dist/machines/test-machines.min.js
360K	dist/machines/test-machines.min.js.map

MASTER
[kkoukiou@sequoia cockpit-bak]$ du -sh dist/machines/*
4.0K	dist/machines/index.html
80K	dist/machines/machines.css
100K	dist/machines/machines.css.map
1.4M	dist/machines/machines.min.js
5.1M	dist/machines/machines.min.js.map
148K	dist/machines/Makefile.deps
4.0K	dist/machines/manifest.json
0	dist/machines/stamp
8.0K	dist/machines/test-machines.css
12K	dist/machines/test-machines.css.map
4.0K	dist/machines/test-machines.html
80K	dist/machines/test-machines.min.js
360K	dist/machines/test-machines.min.js.map

```

martinpitt commented:
Thanks for the [standalone example](https://github.com/KKoukiou/patternfly-react-table-example)!

If I build this in development mode, the size is actually comparable to what we get in cockpit:
````diff
index 635148d..ec3511d 100644
--- a/webpack.config.js
+++ b/webpack.config.js
@@ -8,6 +8,7 @@ const htmlPlugin = new HtmlWebPackPlugin({
 });
 
 module.exports = {
+  mode: 'development',
   module: {
     rules: [
       {
```

```
❱❱❱ du -hsc dist/*
8.7M	dist/bundle.js
4.0K	dist/index.html
460K	dist/main.css
9.2M	total
```

It has the same issues with including all SVGs and most/all PF elements (in code and CSS). In `production` mode this indeed gets minified and reduced:

```
❱❱❱ du -hsc dist/*
356K	dist/bundle.js
4.0K	dist/index.html
152K	dist/main.css
512K	total
```
It seems the CSS and JS both have much less unnecessary code. So perhaps webpack just doesn't throw out unused bits in development mode? Thanks for the production size diff of cockpit, that does look reasonable indeed. Sorry for the fuss then, I wasn't aware of this webpack weirdness!

So like that and without the extra loaders this looks great, thanks! I'll trigger tests.

martinpitt commented:
@garrett, do you want to review this, too?

martinpitt commented:
The [selenium](https://209.132.184.41:8493/logs/pull-12573-20190909-080303-4dde95f9-cockpit-project-cockpit--fedora-30-selenium-firefox/log.html) [tests](https://logs-https-cockpit.apps.ci.centos.org/logs/pull-12573-20190909-080303-4dde95f9-cockpit-project-cockpit--fedora-30-selenium-chrome/log.html) repeatedly fail, which is suspicious -- they might need adjustment to the changed CSS here?

KKoukiou commented:
> The [selenium](https://209.132.184.41:8493/logs/pull-12573-20190909-080303-4dde95f9-cockpit-project-cockpit--fedora-30-selenium-firefox/log.html) [tests](https://logs-https-cockpit.apps.ci.centos.org/logs/pull-12573-20190909-080303-4dde95f9-cockpit-project-cockpit--fedora-30-selenium-chrome/log.html) repeatedly fail, which is suspicious -- they might need adjustment to the changed CSS here?

Hmm, the chrome machines tests are all green,
the firefox fail with some blank page as screeshot, so I guess selenium server issue, https://209.132.184.41:8493/logs/pull-12573-20190909-080303-4dde95f9-cockpit-project-cockpit--fedora-30-selenium-firefox/screenshot-testCreate20VMs-646597.png 

And then the edge are the most weird, since they clearly show an oops in the main Virtual Machines page. Checking this now.

martinpitt commented:
Whee! Welcome, PatternFly 4!
